li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaInstrumentedConsumerImpl.java

```diff
@@ -0,0 +1,583 @@
+  }
+
+  @Override
+  public void resume(Collection<TopicPartition> partitions) {
+    verifyOpen();
+    delegate.resume(partitions);
+  }
+
+  @Override
+  public Map<TopicPartition, OffsetAndTimestamp> offsetsForTimes(Map<TopicPartition, Long> timestampsToSearch) {
+    verifyOpen();
+    return delegate.offsetsForTimes(timestampsToSearch);
+  }
+
+  @Override
+  public Map<TopicPartition, OffsetAndTimestamp> offsetsForTimes(Map<TopicPartition, Long> timestampsToSearch,
+      Duration timeout) {
+    verifyOpen();
+    return delegate.offsetsForTimes(timestampsToSearch, timeout);
+  }
+
+  @Override
+  public Map<TopicPartition, Long> beginningOffsets(Collection<TopicPartition> partitions) {
+    verifyOpen();
+    return delegate.beginningOffsets(partitions);
+  }
+
+  @Override
+  public Map<TopicPartition, Long> beginningOffsets(Collection<TopicPartition> partitions, Duration timeout) {
+    verifyOpen();
+    return delegate.beginningOffsets(partitions, timeout);
+  }
+
+  @Override
+  public Map<TopicPartition, Long> endOffsets(Collection<TopicPartition> partitions) {
+    verifyOpen();
+    return delegate.endOffsets(partitions);
+  }
+
+  @Override
+  public Map<TopicPartition, Long> endOffsets(Collection<TopicPartition> partitions, Duration timeout) {
+    verifyOpen();
+    return delegate.endOffsets(partitions, timeout);
+  }
+
+  @Override
+  public void close() {
+    if (!proceedClosing()) {
+      return;
+    }
+    try {
+      delegate.close();
+    } finally {
+      delegate = null;
+    }
+  }
+
+  @Override
+  @Deprecated
+  public void close(long timeout, TimeUnit unit) {
+    if (!proceedClosing()) {
+      return;
+    }
+    try {
+      delegate.close(timeout, unit);
+    } finally {
+      delegate = null;
+    }
+  }
+
+  @Override
+  public void close(Duration timeout) {
+    if (!proceedClosing()) {
+      return;
+    }
+    try {
+      delegate.close(timeout);
+    } finally {
+      delegate = null;
+    }
+  }
+
+  @Override
+  public void wakeup() {
+    //vanilla allows this even if closed
+    Consumer<K, V> delegate = this.delegate;
+    if (delegate != null) {
+      delegate.wakeup();
+    }
+  }
+
+  private boolean proceedClosing() {
+    if (isClosed()) {
+      return false;
+    }
+    delegateLock.writeLock().lock();
+    try {
+      if (isClosed()) {
+        return false;
+      }
+      closedAt = System.currentTimeMillis();
+      return true;
+    } finally {
+      delegateLock.writeLock().unlock();
+    }
+  }
+
+  private boolean isClosed() {
+    return closedAt != null;
+  }
+
+  /**
+   * throws an exception if this consumer has been closed.
+   * there's also an assumption that unless closed, delegate is always != null
+   */
+  private void verifyOpen() {
+    if (isClosed()) {
+      //same exception thats thrown by a vanilla consumer
+      throw new IllegalStateException("This consumer has already been closed.");
+    }
+  }
+
+  private class MetricsProxy implements Map {
```

Comment:
Why do we need MetricsProxy, instead of simply implementing metrics() as getDelegete() below?

Reply:
users can get a reference to the sensor map and hold onto it beyond the life of the delegate client it came from.